### PR TITLE
[ENH] Added FitInTransform transformer

### PR DESCRIPTION
**Pull request by aiwalter**
#### Reference Issues/PRs
#2267 #2382 

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.
- [ ] The PR title starts with either [ENH], [DOC] or [BUG] indicating whether the PR topic is related to enhancement, documentation or bug

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Review by fkiraly — COMMENTED** [review]
I'd put it in `transformations.compose` where the pipelines are, what do you think?

File: sktime/transformations/series/compose.py
Comment (fkiraly) [inline sktime/transformations/series/compose.py:546]: I think `update` should be empty. The idea is for this to be a "stateless" version that is always applied to the batch.

File: sktime/transformations/series/compose.py
Comment (fkiraly) [inline sktime/transformations/series/compose.py:508]: this will fail the "transform does not change state" test.

The way to avoid this would be to not write to `self` at all, just remove the `transformer_` attribute, and create the clone locally in `_transform`.

File: sktime/transformations/series/compose.py
Comment (fkiraly) [inline sktime/transformations/series/compose.py:12]: estimators used only in the test params should be imported inside `get_test_params`

**Comment by aiwalter** [comment]
> I'd put it in transformations.compose where the pipelines are, what do you think?

hm but then e.g. `OptionalPassthough` should be there as well? Do we actually still need the `series` and `panel` modules at all?

File: sktime/transformations/series/compose.py
Comment (aiwalter) [reply sktime/transformations/series/compose.py:508]: but I need to write to `self` in order to enable `inverse_transform` 🤔

Comment (fkiraly) [reply sktime/transformations/series/compose.py:508]: would `inverse_transform` not call `fit.inverse_transform`?

**Comment by fkiraly** [comment]
> hm but then e.g. OptionalPassthough should be there as well? Do we actually still need the series and panel modules at all?

I think the division still makes sense, but it hasn't been really adhered to correctly. So I was considering removing it, but we haven't decided on it.

Either way, I think all compositors should go in `transformations.compose`, at least the new ones.
I was going to move the `OptionalPassthrough` there, too.

File: sktime/transformations/series/compose.py
Comment (aiwalter) [reply sktime/transformations/series/compose.py:508]: what is `fit` here?

**Comment by aiwalter** [comment]
also tests are passing? So state change seems to be fine?

File: sktime/transformations/series/compose.py
Comment (fkiraly) [reply sktime/transformations/series/compose.py]: of the inner trafo

**Comment by fkiraly** [comment]
>also tests are passing? So state change seems to be fine?

No, you need to test with an inner transformer where not `fit_is_empty`, otherwise the wrapper does nothing...

File: sktime/transformations/tests/test_fitintransform.py
Comment (fkiraly) [inline sktime/transformations/tests/test_fitintransform.py:22]: same here, you want to use the boxcox or similar transform that does not have empty `_fit`.
why not use `create_test_instance`?

**Comment by fkiraly** [comment]
It appears that the tests were turned off since there were some estimators make state changes in `predict`, `transform`.

That is "cheating", I guess we turned it off in a refactor to the more strict interface, and didn't properly turn it on again.

Can you kindly ensure nothing writes to `self` or changes attributes of `self` in `transform`, while I fix the tests?

**Comment by aiwalter** [comment]
I need to write to `self`, otherwise I  cant do `inverse_trasnform`. Also why is it actually not allowed to write to `self`?

**Comment by fkiraly** [comment]
>I need to write to `self`, otherwise I  cant do `inverse_transform`. 

Can you explain that? I thought `inverse_transform` only requires `fit`, but not `transform` to have been run.

> Also why is it actually not allowed to write to `self`?

Because the state change design for all estimators allows state change only through `fit` and `update`.
This is also written down in the base class docstrings, and all estimators must adhere to that.

The entire logic is based around that assumption, so we can't change it on an ad-hoc basis.
It's one of the "core interface assumptions".

If you are asking for the rationale behind that:

* sklearn `BaseEstimator` interface makes the same assumptions
* it makes sense to localize state change to only a small number of methods. The more methods change state, the more possibilities you have to be mindful of in transitions. Worst case, you get combinatorial explosion of state change sequences which is really hard to control, test, and explain to a user.

**Comment by aiwalter** [comment]
> Can you explain that? I thought inverse_transform only requires fit, but not transform to have been run.

well how can I do `self.forecaster.inverse_transform()` if this fitted `forecaster_` is not soted? If I just to `return self._forecaster_.fit_transform(X, y)` in `_transform()`, then I dont store this fit and then I cant do `inverse_transform`

**Comment by fkiraly** [comment]
> well how can I do `self.forecaster.inverse_transform()` if this fitted `forecaster_` is not stored? If I just to `return self._forecaster_.fit_transform(X, y)` in `_transform()`, then I dont store this fit and then I cant do `inverse_transform` 

Why does `fit(X, y).inverse_transform(X, y)` not work?

File: sktime/transformations/tests/test_fitintransform.py
Comment (fkiraly) [inline sktime/transformations/tests/test_fitintransform.py:4]: this should say `FitInTransform`

File: sktime/transformations/compose.py
Comment (fkiraly) [inline sktime/transformations/compose.py:705]: I´d recommend a `ClassifierPipeline` becaues there the algorithm actually makes sense...

File: sktime/transformations/compose.py
Comment (fkiraly) [inline sktime/transformations/compose.py:672]: I would recommend to be more precise here, along the following lines. It´s not about "accuracy", but you may want to apply the transformation to the same sample. This can make sense in classification, but usually doesn´t make sense in forecasting (since it can lead to information leakage).

File: sktime/transformations/compose.py
Comment (fkiraly) [inline sktime/transformations/compose.py:670]: typo

**Review by fkiraly — APPROVED** [review]
Ok, good for me.

I´ve left some non-blocking, docstring related comments.

File: sktime/transformations/compose.py
Comment (aiwalter) [reply sktime/transformations/compose.py:705]: sure we can add this as well, can you make a proposal? I am not so familiar with the classification module

Comment (fkiraly) [reply sktime/transformations/compose.py:705]: `pipe = imputer * SummaryClassifier()`

Comment (aiwalter) [reply sktime/transformations/compose.py:705]: I think we shouldnt use dunder methods in examples, if s.o. sees it first time it just looks confusing

Comment (fkiraly) [reply sktime/transformations/compose.py:705]: so just pick one of the examples from `ClassificationPipeline` and replace the transformer

**Comment by fkiraly** [comment]
@aiwalter, do you want to fix the docstrings before we merge?

File: sktime/transformations/compose.py
Comment (aiwalter) [reply sktime/transformations/compose.py:705]: i dont find a class `ClassificationPipeline`?

File: sktime/transformations/compose.py
Comment (aiwalter) [reply sktime/transformations/compose.py:672]: pls provide description you want to have

**Comment by aiwalter** [comment]
its ready to be merged

**Comment by fkiraly** [comment]
tests seem stuck, let´s restart

File: sktime/transformations/compose.py
Comment (fkiraly) [inline sktime/transformations/compose.py:672]:
```suggestion
    In panel settings, e.g., time series classification, it can be preferable (or, necessary) to fit and transform on the test set, e.g., interpolate within the same series that interpolation parameters are being fitted on.
    `FitInTransform` can be used to wrap any transformer to ensure that `fit` and `transform` happen always on the same series, by delaying the `fit` to the `transform` batch.
    Warning: use of `FitInTransform` will typically not be useful, or can constitute a mistake (data leakage) when naively used in a forecasting setting.
```

File: sktime/transformations/compose.py
Comment (fkiraly) [reply sktime/transformations/compose.py:672]: done

File: sktime/transformations/compose.py
Comment (fkiraly) [inline sktime/transformations/compose.py:680]: this looks like a formatting accident??

**Review by fkiraly — DISMISSED** [review]
Happy to merge.

Looks like there is a formatting accident in the docstring, but happy to merge (and possibly fix) once tests pass again.

**Pull request merged**
